### ISSUE TITLE
Figure out the best way to handle actually installing the role in scenarios

### ISSUE THREAD

**Issue by tehsmyers** [oasis-roles/molecule_openstack_ci#2]
Since molecule runs `destroy` before `dependency`, `destroy` fails due to the missing dependency. I have a few solutions.

## First Solution

This one's a little annoying, but is the easiest to maintain. Add an option to oasis-pipeline to install dependencies before running `molecule test`, default it to on. Add a note to the README on this role telling you to `molecule dependency -s <scenario>` before provisioning.

## Second solution

This one's more annoying and (barely) harder to maintain, but requires no changes to oasis-pipeline and has no potential for causing unexpected outcomes. In the molecule config, explicitly place `dependency` before `destroy` where appropriate:

```yaml
# somewhere in molecule.yml, for example
check_sequence: ['dependency', 'destroy', 'create', 'prepare', 'converge', 'check', 'destroy']
test_sequence: ['dependency', 'lint', 'destroy', 'syntax', 'create', 'prepare', 'converge', 'idempotence', 'side_effect', 'verify', 'destroy']
```

## Third solution

Add tasks to install this role to the example `create.yml` and `destroy.yml`. This would work great, but feels kinda dirty?

## Upstream solution

Not exclusive of the other options, we could open an issue with molecule to gauge interest on putting the dependency step before destroy by default for the explicit purpose of using roles like this to simplify using molecule. If I were a molecule dev, I'd be really wary of making a change like this since it seemingly arbitrarily could break existing workflows, but the idea of using roles to make using molecule easier might be novel and enticing enough to warrant such a change.

## Other solutions

Stuff I haven't thought of...

**Comment by greg-hellings** [comment]
> Since molecule runs `destroy` before `dependency`, `destroy` fails due to the missing dependency. I have a few solutions.
> ## First Solution
> 
> This one's a little annoying, but is the easiest to maintain. Add an option to oasis-pipeline to install dependencies before running `molecule test`, default it to on. Add a note to the README on this role telling you to `molecule dependency -s <scenario>` before provisioning.

All this will do is increase time. In the most non-Ansible-like thing that Molecule could ever do, they force the installation of dependencies every time. So we'd still be getting dependencies re-installed again. I'm not down with this.

> ## Second solution
> 
> This one's more annoying and (barely) harder to maintain, but requires no changes to oasis-pipeline and has no potential for causing unexpected outcomes. In the molecule config, explicitly place `dependency` before `destroy` where appropriate:
> 
> ```yaml
> # somewhere in molecule.yml, for example
> check_sequence: ['dependency', 'destroy', 'create', 'prepare', 'converge', 'check', 'destroy']
> test_sequence: ['dependency', 'lint', 'destroy', 'syntax', 'create', 'prepare', 'converge', 'idempotence', 'side_effect', 'verify', 'destroy']
> ```

I have used this in places like https://github.com/oasis-roles/openstack_provision/tree/master/molecule/default/molecule.yml where I need to completely suppress the destroy step (or, at least, where it's completely wasted and therefore should be skipped). I like this solution best, as it's exactly why Molecule gives us the ability to configure this.

> 
> ## Third solution
> 
> Add tasks to install this role to the example `create.yml` and `destroy.yml`. This would work great, but feels kinda dirty?
> ## Upstream solution
> 
> Not exclusive of the other options, we could open an issue with molecule to gauge interest on putting the dependency step before destroy by default for the explicit purpose of using roles like this to simplify using molecule. If I were a molecule dev, I'd be really wary of making a change like this since it seemingly arbitrarily could break existing workflows, but the idea of using roles to make using molecule easier might be novel and enticing enough to warrant such a change.

I'm not saying you're wrong, but I can't imagine how it would damage the behavior of Molecule. While we're proposing changes to their defaults, I would also like to propose that the `create` step include `dependency` by default, since I often use roles in the `prepare` playbook. When you execute `create` it also executes `prepare` if the host has not been prepared yet. As such, it would be great if it executed `dependency` first, but alas it does not.

I don't know how they'll feel about feedback like that, but I don't see any reason that we shouldn't propose it, at least. The worst that happens is they say "no".

> ## Other solutions
> 
> Stuff I haven't thought of...

So my vote is for solution 2 until and pursue an upstream proposal to increase/change the executions of the `dependency` step.

**Comment by tehsmyers** [comment]
Both the first and second solution work well enough that I'll throw them in the readme, at least. I'll probably implement this in rhsm using the first solution in a pre-test hook.

**Comment by tehsmyers** [comment]
> All this will do is increase time. ...

Yeah, that's a really good point.

**Comment by tehsmyers** [comment]
## Fourth Solution

oasis-pipeline already supports installing roles from a requirements.yml at the root of a role repository. So...add a requirements.yml at the root of a role repository to ensure the role is installed before running molecule. Add a note to OASIS developer docs that if requirements.yml exists at the root of a repository, you should `ansible-galaxy install` it before you develop.

**Comment by tehsmyers** [comment]
Ah, I was wondering why I didn't suggest that fourth solution in the original post. I completely forgot just now that roles installed with ansible-galaxy install are not installed in a place where molecule will look at them, for isolation purposes. It's likely that having support for requirements.yml in oasis-pipeline is entirely useless. :)

**Comment by greg-hellings** [comment]
I just was sitting here dreaming about this, and I think that Solution 1 is superior. Just add pipeline code to run dependency. Solution 2 becomes a nightmare if we ever upgrade Molecule (which I hope we can do) and they ever add, move, remove, or change the order of sequences. Or if they add extra steps that we're now skipping because we are forcing this.

**Comment by tehsmyers** [comment]
I'll go with Solution 1 for now, but this is also worth mentioning:
https://github.com/ansible/molecule/issues/1734

**Comment by tehsmyers** [comment]
I've got a PR up on [oasis-pipeline](https://github.com/oasis-origin/oasis-pipeline/pull/7), but I've also got a PR up on molecule itself related to the issue I linked in my previous comment. I'll get the pipeline PR fixed up and tested, but depending on the timeline for getting that merged and the next molecule release, it may be a short-lived workaround.

**Comment by greg-hellings** [comment]
So I've just encountered a complication in running this as proposed: it appears the "destroy" sequence and command cleans up all the installed dependency roles. When I customize my process to include dependency before destroy in a run of `molecule test`, then destroy runs successfully the first time but create fails (leading to a second run of destroy which also fails). Slotting dependency in both before AND between destroy/create leads to success. That is, the following succeeds:

```yaml
scenario:
  name: default
  destroy_sequence:
    - dependency
    - destroy
  test_sequence:
    - lint
    - dependency
    - syntax
    - destroy
    - dependency # Without this, there is great agony and consternation
    - create
    - prepare
    - converge
    - idempotence
    - side_effect
    - verify
    - destroy
```

Not the best, but it's what I had to do.

**Comment by tehsmyers** [comment]
Oh my.

As a workaround, you can `ansible-galaxy install -p .. oasis_roles.molecule_openstack_ci`, since apparently molecule adds the parent directory to the roles path. This probably works fine for jenkins but it not automatic and will junk up developer envs.

I was able to reproduce what you're seeing, but I have yet to find the code that actually removes the roles from the ephemeral dir on destroy.

**Comment by tehsmyers** [comment]
Found it: https://github.com/seandst/molecule/blob/master/molecule/command/base.py#L59-L80

**Comment by greg-hellings** [comment]
An alternative proposal in your bug report/PR would be to introduce *yet another* step in the pipeline. It would be something like provision_dependency and would look for provision_requirements.yml in the scenario directory. Upon location of that file, it would run that prior to any destroy/create/prepare steps.

Personally, I like that idea as I am unhappy having to wait for roles to download *yet again* each time I try to do something if those roles are only used in the (create,destroy,prepare).yml steps of the pipeline. It makes it very clear that certain roles are only necessary for some functionality and are not explicit dependencies of the role or called out of it, etc.

I'm very happy to be completely wrong on that if I'm barking up the wrong creek.

**Comment by tehsmyers** [comment]
There are open bugs about the dependency installation working that way that I was hoping to avoid getting sucked into, which is why I've (so far) stayed away from adding another provision step. I think that adding a `provision_dependency` step is reasonable, and could probably reuse the existing dependency behavior, but would still run into this purge problem because of how the behavior is implemented (it happens unconditionally on destroy).

So right now I'm looking at taking that purge behavior out of the destroy step and moving it somewhere else, like up to the overall scenario level. Then I'll add an option like `--purge [on_destroy|never]` to mirror the existing `--destroy [always|never]`, defaulting to the current effective `--purge on_destroy` behavior. Then probably I'll get sucked into the dependency tickets and add config/cmdline flags for how to give scenario writers more control over when dependencies are redownloaded. You know this, but I'll say it anyway: There are valid justifications for keeping the installed dependency roles around by default for developers, there are valid justifications for always redownloading in CI, and I think it's possible to appease both audiences by shuffling some things around in molecule and exposing some control over its currently hard-coded behavior.

**Comment by tehsmyers** [comment]
Been digging into molecule a little bit more, and working with current molecule developers, and the short story right now is that nothing suggested so far will work because of how molecule works. A modification of the provision dependency step might work, though, so I'm currently exploring how we might augment the molecule provisioners to be able to do this without necessarily having to add a new step, or otherwise alter the scenario sequence execution to be a little smarter about how/when it "prunes" the ephemeral dir.

**Comment by tehsmyers** [comment]
Work has continued on this in the molecule project itself. The approach in my previous comment, which was to alter molecule's default scenario sequence and how/when it decides to prune the ephemeral dir, was accepted in https://github.com/ansible/molecule/issues/1734 and should be included in the upcoming molecule 2.21 release.

**Comment by tehsmyers** [comment]
Since this issue is "figure out the best way...", and that's done now, I think this issue can be closed.
